### PR TITLE
KAFKA-9481: Graceful handling TaskMigrated and TaskCorrupted

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Removed task field from TaskMigrated; the only caller that encodes a task id from StreamTask actually do not throw so we only log it. To handle it on StreamThread we just always enforce rebalance (and we would call onPartitionsLost to remove all tasks as dirty).

2. Added TaskCorruptedException with a set of task-ids. The first scenario of this is the restoreConsumer.poll which throws InvalidOffset indicating that the logs are truncated / compacted. To handle it on StreamThread we first close the corresponding tasks as dirty (if EOS is enabled we would also wipe out the state stores), and then revive them into the CREATED state.

3. Also fixed a bug while investigating KAFKA-9572: when suspending / closing a restoring task we should not commit the new offsets but only updating the checkpoint file.

4. Re-enabled the unit test.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
test this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 1]
```diff
@@ -415,6 +415,8 @@ public void close() throws ProcessorStateException {
                     log.error("Failed to close state store {}: ", store.name(), exception);
                 }
             }
+
+            stores.clear();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:438]: We need to clear the stores map now since we may re-initialize the state stores upon reviving a task.

File: streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+package org.apache.kafka.streams.errors;
+
+import org.apache.kafka.streams.processor.TaskId;
+
+import java.util.Set;
+
+/**
+ * Indicates a specific task is corrupted and need to be re-initialized. It can be thrown when
+ *
+ * 1) Under EOS, if the checkpoint file does not contain offsets for corresponding store's changelogs, meaning
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java:28]: This case 1) would be done in another PR, I just added the java-doc here to complete the scope.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java [hunk 3]
```diff
@@ -690,18 +704,18 @@ private void pauseChangelogsFromRestoreConsumer(final Collection<TopicPartition>
...
     }
 
     private void removeChangelogsFromRestoreConsumer(final Collection<TopicPartition> partitions) {
         final Set<TopicPartition> assignment = new HashSet<>(restoreConsumer.assignment());
 
-        // the current assignment should contain the all partitions to remove
-        if (!assignment.containsAll(partitions)) {
-            throw new IllegalStateException("The current assignment " + assignment + " " +
-                "does not contain some of the partitions " + partitions + " for removing.");
+        if (assignment.removeAll(partitions)) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java]: Here I made the `remove` call idempotent.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java [hunk 4]
```diff
@@ -154,6 +154,11 @@ public boolean isValidTransition(final State newState) {
      */
     void closeDirty();
 
+    /**
+     * Revive a closed task to a created one; should never throw an exception
+     */
+    void revive();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java:160]: Okay this might be a bit controversial: I tried to make the re-initialization logic as cheap as possible since otherwise we may be kicked out of the group for not calling consumer.poll in time. After looking at the source code I found the only part is the closure of the task-level sensors which I moved out of the `task.close` functions. So from CLOSED -> CREATED there's nothing necessary.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 5]
```diff
@@ -117,6 +120,19 @@ void handleRebalanceComplete() {
         rebalanceInProgress = false;
     }
 
+    void handleCorruption(final Set<TaskId> taskIds) {
+        for (final TaskId taskId : taskIds) {
+            final Task task = tasks.get(taskId);
+
+            // this call is idempotent so even if the task is only CREATED we can still call it
+            changelogReader.remove(task.changelogPartitions());
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:130]: We have to call changelog.remove() before task.closeDirty since now we clear the `stores` map in closeDirty, and after that task.changelogPartitions() would return nothing.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 6]
```diff
@@ -147,20 +163,18 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
                 task.resume();
                 standbyTasksToCreate.remove(task.id());
             } else /* we previously owned this task, and we don't have it anymore, or it has changed active/standby state */ {
-                final Set<TopicPartition> inputPartitions = task.inputPartitions();
+                cleanupTask(task);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:177]: Consolidated a couple of the same pattern into this private function.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 7]
```diff
@@ -285,18 +299,13 @@ void handleLostAll() {
...
-            final Set<TopicPartition> inputPartitions = task.inputPartitions();
             // Even though we've apparently dropped out of the group, we can continue safely to maintain our
             // standby tasks while we rejoin.
             if (task.isActive()) {
+                cleanupTask(task);
                 task.closeDirty();
-                changelogReader.remove(task.changelogPartitions());
-            }
-
-            for (final TopicPartition inputPartition : inputPartitions) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:304]: Not sure why we do not remove the input partitions previously.. is that intentional @vvcephei ? If not I'd move the removal into the block (I already did just to clarify :P)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 8]
```diff
@@ -331,12 +340,27 @@ void handleLostAll() {
         return locallyStoredTasks;
     }
 
+    private void cleanupTask(final Task task) {
+        // 1. remove the changelog partitions from changelog reader;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:362]: The order here cannot be changed so I left this comment.

**Review by guozhangwang — COMMENTED** [review]
@vvcephei @ableegoldman

**Comment by guozhangwang** [comment]
The unit test suite passed locally (Java8, Scala 2.12). Currently I cannot trigger a jenkins build out of it (not knowing why).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java [hunk 9]
```diff
@@ -70,6 +71,15 @@ public boolean isClosed() {
         return state;
     }
 
+    @Override
+    public void revive() {
+        if (state == CLOSED) {
+            transitionTo(CREATED);
+        } else {
+            throw new IllegalStateException("Illegal state " + state() + " while committing standby task " + id);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java]: Remove "standby" from error message (unless this only applies to standbys?)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:438]: Do we also need to clear `storeToChangelogTopic`, etc?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java [hunk 10]
```diff
@@ -258,6 +260,8 @@ private boolean hasRestoredToEnd(final ChangelogMetadata metadata) {
                 // if we cannot get the position of the consumer within timeout, just return false
                 return false;
             } catch (final KafkaException e) {
+                // this also includes InvalidOffsetException, which should not happen under normal
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:261]: Just wondering, why is it ok to get `InvalidOffsetException` during `restore`/`poll` but not here? When might this get thrown from `#position`? ditto for in `prepareChangelogs` down below

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 11]
```diff
@@ -692,9 +688,10 @@ private void closeRecordCollector(final boolean clean) {
     @Override
     public void addRecords(final TopicPartition partition, final Iterable<ConsumerRecord<byte[], byte[]>> records) {
         if (state() == State.CLOSED || state() == State.CLOSING) {
-            log.info("Stream task {} is already closed, probably because it got unexpectedly migrated to another thread already. " +
-                         "Notifying the thread to trigger a new rebalance immediately.", id());
-            throw new TaskMigratedException(id());
+            // a task is only closing / closed when 1) task manager is closing, 2) a rebalance is undergoing;
+            // in either case we can just log it and move on without notifying the thread since the consumer
+            // would soon be updated to not return any records for this task anymore.
+            log.info("Stream task {} is already in {} state, skip adding records to it.", id(), state());
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Hm. I think we should actually be concerned if we ever get to here -- I'm not sure the `TaskMigratedException` made sense either. Trying to add records to a closed task would imply that we closed the task due to shutting down or because we no longer own it, both cases which should also involves trimming its topic(s) from the consumer's assignment, but were still returned records for said topic(s).

Unless, the consumer may still return already-fetched records from partitions no longer in its assignment during poll? I thought we would trim those records out and only return from the actual assignment

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 12]
```diff
@@ -745,14 +752,20 @@ private void runLoop() {
...
-                             "Will try to rejoin the consumer group.", ignoreAndRejoinGroup.migratedTaskId());
+            } catch (final TaskCorruptedException e) {
+                log.warn("Detected the states of tasks {} are corrupted. " +
+                    "Will close the task as dirty and re-create and bootstrap from scratch.", e.corruptedTaskIds());
+
+                taskManager.handleCorruption(e.corruptedTaskIds());
+            } catch (final TaskMigratedException e) {
+                log.warn("Detected that the thread is being fenced. " +
+                    "This implies that this thread missed a rebalance and dropped out of the consumer group. " +
+                    "Will migrate out all assigned tasks and rejoin the consumer group.");
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]:
```suggestion
                    "Will close out all assigned tasks and rejoin the consumer group.");
```

**Review by ableegoldman — COMMENTED** [review]
Still need to look through TaskManager, but left a few initial comments on the rest

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:438]: `storeToChangelogTopic` and `sourcePartitions` are passed in at construction time and final, so we cannot clear them (since they would only be initialized once).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java (hunk 10 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:261]: `consumer.poll` throwing `InvalidOffsetException` should be handled as TaskCorrupted; `consumer.position` throwing `InvalidOffsetException` should not happen under normal scenarios, when it happens it indicates a bug and hence we do not need to special handle it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 11 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Here's my rationale:

If the task is closed due to rebalance (i.e. we `#handleAssignment` or `#handleLostAll`), there might still be some buffered records from the consumer that are returning (since we update the subscription of consumer after), in this case since the subscription would be updated in the next iteration and no records would be returned, it is okay to just skip this once.

If the task is closed due to closing the thread, then there's no need to throw an exception either.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 12 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+ *    previously it was not close cleanly;
+ * 2) Out-of-range exception thrown during restoration, meaning that the changelog has been modified and we re-bootstrap
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java:30]: Just now having this thought... Supposing this happens, is it guaranteed to apply to all the stores in the task? I.e., do we really need to re-bootstrap all the stores, or just the one(s) for which our offset is out of range?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java [hunk 14]
```diff
@@ -409,6 +413,15 @@ public void restore() {
             } catch (final FencedInstanceIdException e) {
                 // when the consumer gets fenced, all its tasks should be migrated
                 throw new TaskMigratedException("Restore consumer get fenced by instance-id polling records.", e);
+            } catch (final InvalidOffsetException e) {
+                log.warn("Encountered {} fetching records from restore consumer for partitions {}, " +
+                    "marking the corresponding tasks as corrupted.", e.getClass().getName(), e.partitions());
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java]: I guess it wouldn't hurt to explain what the exception means (our position is too old and has been deleted or compacted by the broker) and what we hope to accomplish by marking the task as corrupted (to re-bootstrap the stores from the changelog and return to normal processing).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 7 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:304]: The intent was to remove the input partitions from the map any time we remove a task from `tasks`. It looks like your code maintains this (in a clearer and cleaner way).

**Review by vvcephei — COMMENTED** [review]
Thanks for this, @guozhangwang ! Comments below.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 7 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:304]: Got it. The previous code is that we would `partitionToTask.remove(inputPartition);` and remove the task no matter if the task is closed or not, which is a bit weird --- for standby tasks, we do not close them, but we still remove from the iterator and remove from the materialized `partitionToTask`. My modification is to ONLY do this logic if we are closing the task.

As long as you agree this is correct I'm relieved.

File: streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java:30]: Yes, we can re-bootstrap the state stores only -- this is what we did in the past but that was a lot messier (remember we have to use the optional in fixed-order map? :P). My thoughts are that for non-EOS, the checkpoint file would likely exist so even re-bootstrap the whole task would be okay, for EOS, it is safer to re-bootstrap the whole task.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java (hunk 14 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java]: Ack.

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java (hunk 13 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java:30]: It's certainly safer, but the performance hit seems concerning... restoration i/o is already one of the things people complain about most, and this choice could amplify it multiple times over.

Maybe we can handle it more cleanly by closing all the stores nicely, writing a checkpoint file with the out-of-range stores' checkpoints at 0, and then re-bootstrapping the task, so it only has to restore the broken stores?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 7 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:304]: Ah, then it was my mistake before! Good catch.

File: streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/errors/TaskCorruptedException.java:30]: That sounds a good idea. Let me try it out.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 15]
```diff
@@ -436,6 +436,15 @@ private void close(final boolean clean) {
         transitionTo(State.CLOSED);
     }
 
+    @Override
+    public void markChangelogAsCorrupted(final Set<TopicPartition> partitions) {
+        stateMgr.markChangelogAsCorrupted(partitions);
+
+        // only write a new checkpoint (excluding the corrupted partitions) if eos is disabled
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: We can checkpoint (excluding the corrupted partitions) for 1) standby tasks and 2) non-eos active tasks, for eos active tasks we should not write the checkpoint since for eos we HAVE TO reboot every store from scratch to maintain consistency.

**Review by guozhangwang — COMMENTED** [review]
@vvcephei I made the change and it's a bit more complicated than I thought, the current proposal is the smallest change that I can make (it still added new APIs into Task interface). LMK WDYT.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 16]
```diff
@@ -460,6 +460,15 @@ private void close(final boolean clean) {
      * source topic partitions, or if it is enforced to be processable
      */
     public boolean isProcessable(final long wallClockTime) {
+        if (state() == State.CLOSED || state() == State.CLOSING) {
+            // a task is only closing / closed when 1) task manager is closing, 2) a rebalance is undergoing;
+            // in either case we can just log it and move on without notifying the thread since the consumer
+            // would soon be updated to not return any records for this task anymore.
+            log.info("Stream task {} is already in {} state, skip adding records to it.", id(), state());
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Is this what you meant?

```suggestion
            log.info("Stream task {} is already in {} state, skip processing it.", id(), state());
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 17]
```diff
@@ -706,12 +715,6 @@ private void closeRecordCollector(final boolean clean) {
      */
     @Override
     public void addRecords(final TopicPartition partition, final Iterable<ConsumerRecord<byte[], byte[]>> records) {
-        if (state() == State.CLOSED || state() == State.CLOSING) {
-            log.info("Stream task {} is already closed, probably because it got unexpectedly migrated to another thread already. " +
-                         "Notifying the thread to trigger a new rebalance immediately.", id());
-            throw new TaskMigratedException(id());
-        }
-
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:727]: Should we still skip adding records? It looks like that was the intent, but I think what actually would happen is that we'd still add the records, but skip processing them.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 18]
```diff
@@ -718,14 +719,11 @@ public void run() {
...
-        } catch (final KafkaException e) {
-            log.error("Encountered the following unexpected Kafka exception during processing, " +
-                          "this usually indicate Streams internal errors:", e);
-            throw e;
         } catch (final Exception e) {
             // we have caught all Kafka related exceptions, and other runtime exceptions
             // should be due to user application errors
-            log.error("Encountered the following error during processing:", e);
+            log.error("Encountered the following exception during processing " +
+                "and the thread is going to shutdown: ", e);
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]:
```suggestion
                "and the thread is going to shut down: ", e);
```

I'm a nerd, I know... "shutdown" is a noun, "shut down" is a verb phrase ("down" modifies the verb "shut").

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 19]
```diff
@@ -745,15 +743,22 @@ private void runLoop() {
             try {
                 runOnce();
                 if (assignmentErrorCode.get() == AssignorError.VERSION_PROBING.code()) {
-                    log.info("Version probing detected. Triggering new rebalance.");
+                    log.info("Version probing detected. Rejoin the consumer group to trigger a new rebalance.");
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]:
```suggestion
                    log.info("Version probing detected. Rejoining the consumer group to trigger a new rebalance.");
```

It sounded like you were telling the operator to do something, but you're just telling them that something is going to happen.

**Review by vvcephei — COMMENTED** [review]
Thanks @guozhangwang , I had a few nits, but the main concern is about the skipping-adding-records thing.

Otherwise, this change looks great to me!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 17 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:727]: This is intentional: as in this PR https://github.com/apache/kafka/pull/8091 where we fixed the committing offset we would look into the buffered record so that we can get the correct "next" offset to commit, if we skip adding records here when the task is closing we would return incorrect results potentially.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 20]
```diff
@@ -239,25 +236,32 @@ public void suspend() {
         if (state() == State.CREATED || state() == State.CLOSING || state() == State.SUSPENDED) {
             // do nothing
             log.trace("Skip suspending since state is {}", state());
+        } else if (state() == State.RUNNING) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:239]: Here is the attempted fix of https://issues.apache.org/jira/browse/KAFKA-9572: if we are closing / suspending a restoring task, we should only update the checkpoint file but should NOT commit offsets, since the committed offsets indicate the  "restore end" and should not be updated, cc @cadonna who filed the JIRA.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 21]
```diff
@@ -413,26 +417,23 @@ private void close(final boolean clean) {
...
-                        stateMgr.flush();
-                    } catch (final RuntimeException error) {
-                        log.debug("Ignoring flush error in unclean close.", error);
-                    }
+                } else {
+                    executeAndMaybeSwallow(false, stateMgr::flush, "state manager flush");
                 }
 
+                transitionTo(State.CLOSING);
+            } else if (state() == State.RESTORING) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:431]: This is part of the fix as well: only flushing / checkpointing, but not committing.

**Review by guozhangwang — COMMENTED** [review]
Adding the fix for KAFKA-9572

**Review by vvcephei — APPROVED** [review]
Thanks @guozhangwang !

**Comment by guozhangwang** [comment]
The local run in JDK8 / Scala 2.12 passed. Will merge to trunk now.
